Aggregate by day

ahaywardtvuk opened the pull request:
This adds an API parameter to the accounting usage data which requests that data is time aggregated by day or month.

This should allow the accounting UI to show more useful and easily navigable data. The list is otherwise extremely long and entries cover variable lengths of time.

ahaywardtvuk committed:
Allow aggregating by day or month when fetching billing data for a workspace.

ahaywardtvuk committed:
Support aggregation for retrievals by account.

ahaywardtvuk committed:
Call the time_aggregation parameter time-aggregation instead.

james-hinton commented on accounting_service/models.py:
```diff
@@ -383,7 +396,61 @@ def find_billing_events(
+    FROM selected_bes
+    WINDOW w AS (PARTITION BY datetime(event_start, 'start of {time_aggregation}'), item_id, workspace)
+) AS window_be
+GROUP BY uuid, event_start, event_end, item_id, sku, workspace
```
`sku` isn't actually selected in this query

ahaywardtvuk committed:
Fix the duplication of consumption in aggregated results when the paging split the
results anywhere other than a day boundary.

This switches round the way this is calculated to first select the aggregated events
in a subquery and then filter and page them in an outer query. This makes it easier
to check that the filtering and paging is correct.

ahaywardtvuk committed:
Try to fix 'invalid numeric literal' error which occurs on dev but not in the test harness.

ahaywardtvuk committed:
Try to deal with sqlite vs PostgreSQL type differences

ahaywardtvuk committed:
Try to track down error happening in dev.

ahaywardtvuk committed:
Must log earlier.

ahaywardtvuk committed:
Another go at debugging the difference between the dev and local environment

ahaywardtvuk committed:
Assigning aggregated billing events the lexicographically highest UUID from the
underlying events, rather than the last in the usual order, makes the query
much simpler and should avoid the problem I'm seeing of unstable orders inside
the window function.

ahaywardtvuk committed:
Deal with a sqlite vs postgresql difference

ahaywardtvuk committed:
Return 404 when requesting a page with after=<unknown UUID>.

ahaywardtvuk commented:
I found some more problems with valid `after=` values not being found due to an unstable sort order when windowing, so it now uses the lexicographically last UUID in the window so that the order doesn't matter.

I've also corrected the return status for unknown values of `after` - now 404 instead of 500.

james-hinton commented on accounting_service/app/app.py:
```diff
@@ -260,6 +262,18 @@ def get_workspace_usage_data(
+            alias="time-aggregation",
+            title="Time aggregation of results",
+            description=(
+                "Optionally ggregate usage information into totals for the given time periods - "
```
small typo

james-hinton commented on accounting_service/models.py:
```diff
@@ -328,60 +349,133 @@ def find_billing_events(
+
+            select_aggregated_events = text(
+                f"""
+SELECT {uuid_expr} as uuid,
```
we may want to index a few of these as the system scales (future obviously)

james-hinton commented on accounting_service/models.py:
```diff
@@ -328,60 +349,133 @@ def find_billing_events(
+            ).scalar_one_or_none()
+
+            if after_be is None:
+                raise AfterBillingEventNotFound(f"after={after} not found")
```
Perhaps a little more descriptive e.g. No matching records found for the provided paging `{after}`.

ahaywardtvuk committed:
Fix spelling

ahaywardtvuk committed:
Improve index use when scanning aggregated billing records.

Getting the DB settings and creating the DB engine need to be separated to avoid a
circular import.

It's not clear how big the performance of the index will be as I don't have enough
data for PostgreSQL to even try, but with enable_seqscan=0 the index is used and is
in the correct order for the GROUP BY (eliminating a sort).

Ultimately we may need to aggregate older billing events into day-sized ones in the
DB, but this may be enough for now.
